### PR TITLE
hz

### PR DESCRIPTION
**Pull request by KozhanSerhii**

### CODE REVIEW

File: DrawEngine/DrawEngine.cs [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+        {
+            _salesRepository = new SalesRepository();
+            _booksRepository = new BooksRepository();
+            _salesWorkflow = new SalesWorkflow();
+            _booksWorkflow = new BooksWorkflow();
+        }
+
+        public void PrintMenu()
+        {
+            Console.WriteLine("menu general:");
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: change to native speaker language

File: DrawEngine/DrawEngine.cs [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+            text = "Price";
+            Console.Write(text + new string(' ', dynamicIndent - text.Length) + "|");
+
+            dynamicIndent = lengthSales.maxNumberOfSalesLength; // Динамічний відступ, який можна змінювати
+            text = "Number Of Sales";
+            Console.Write(text + new string(' ', dynamicIndent - text.Length) + "|");
+            Console.WriteLine();
+        }
+        
+        public void PrintMiddleSale(SaleLength lengthSales)
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: PrintMiddleSale -> PrintSalesContent

File: DrawEngine/DrawEngine.cs [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+                Console.Write(text + new string(' ', dynamicIndent - text.Length) + "|");
+
+                dynamicIndent = lengthSales.maxNumberOfSalesLength; // Динамічний відступ, який можна змінювати
+                text = Convert.ToString(sale.Number_Of_Sales); ;
+                Console.Write(text + new string(' ', dynamicIndent - text.Length) + "|");
+
+                Console.WriteLine();
+            }
+        }
+        public void PrintLowSale(SaleLength lengthSales)
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: PrintLowSale -> PrintSaleLowerBorder

File: DrawEngine/DrawEngine.cs [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+            string str = "Book ID";
+            lengthBooks.maxBookIDLength = books.Max(s => s.Book_ID.ToString().Length) > str.Length ? books.Max(s => s.Book_ID.ToString().Length) : str.Length;
+            str = "Title";
+            lengthBooks.maxTitleLength = books.Max(b => b.Title.Length) > str.Length ? books.Max(b => b.Title.Length) : str.Length;
+            str = "Author";
+            lengthBooks.maxAuthorLength = books.Max(b => b.Author.Length) > str.Length ? books.Max(b => b.Author.Length) : str.Length;
+            lengthBooks.maxStringLength = lengthBooks.maxBookIDLength + lengthBooks.maxTitleLength + lengthBooks.maxAuthorLength;
+            return lengthBooks;
+        }
+        public void PrintTopSale(SaleLength lengthSales)
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: PrintTopSale -> PrintSaleColumnsNames

File: DrawEngine/DrawEngine.cs [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+        }
+        public void PrintLowSale(SaleLength lengthSales)
+        {                   
+            for (int i = 0; i <= lengthSales.maxStringLength + 5; i++)
+            {
+                Console.Write("-");
+            }
+            Console.WriteLine();
+        }        
+        public void PrintTopBook(BookLength lengthBooks)
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: PrintTopBook -> PrintBookColumnsNames

File: DrawEngine/DrawEngine.cs [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+            text = "Title";
+            Console.Write(text + new string(' ', dynamicIndent - text.Length) + "|");
+
+            dynamicIndent = lengthBooks.maxAuthorLength; // Динамічний відступ, який можна змінювати
+            text = "Author";
+            Console.Write(text + new string(' ', dynamicIndent - text.Length) + "|");
+            Console.WriteLine();
+        }
+
+        public void PrintMiddleBook(BookLength lengthBooks)
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: PrintBooksContent

File: DrawEngine/DrawEngine.cs [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+
+                dynamicIndent = lengthBooks.maxAuthorLength; // Динамічний відступ, який можна змінювати
+                text = Convert.ToString(book.Author); ;
+                Console.Write(text + new string(' ', dynamicIndent - text.Length) + "|");
+
+                Console.WriteLine();
+            }
+
+        }
+        public void PrintLowBook(BookLength lengthBooks)
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: PrintLowBook -> PrintBookLowerBorder

File: DrawEngine/DrawEngine.cs [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+        {                     
+            for (int i = 0; i <= lengthBooks.maxStringLength + 3; i++)
+            {
+                Console.Write("-");
+            }
+            Console.WriteLine();
+        }
+        public void PrintAllSales()
+        {            
+            PrintTopSale(CalculateSaleLength());
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: call once CalculateSaleLength() and set to variable - then transmit this variable to method as parameter

File: DrawEngine/DrawEngine.cs [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+            PrintMiddleSale(CalculateSaleLength());
+            PrintLowSale(CalculateSaleLength());
+        }
+        public void PrintAllBooks()
+        {            
+            PrintTopBook(CalculateBookLength());
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: call once CalculateBookLength() and set to variable - then transmit this variable to method as parameter

File: Infrastructure/BusinessLogic/BooksWorkflow.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+            {
+                _booksRepository.Add(new Book { Title = dto.Title, Author = dto.Author });
+            }      
+            
+            return true;
+        }
+
+        public bool DeleteEntity(int id)
+        {
+            return (_booksRepository.Remove(id) == true || _salesRepository.Remove(id) == true);
```
Comment (KozhanMykhailo) [inline Infrastructure/BusinessLogic/BooksWorkflow.cs]: _booksRepository.Remove(id) - it is OK
 || _salesRepository.Remove(id) - why do you use logical operator || here ?
 _salesRepository.Remove(id) - there can be a lot of rows in Sale table where BookId == 4 , but Id of sale can be 4,7,567,56676 egt - how to delete all these rows if they have the same BookId ?

File: DrawEngine/DrawEngine.cs (hunk 1 above)
Comment (KozhanMykhailo) [reply DrawEngine/DrawEngine.cs]: done

File: DrawEngine/DrawEngine.cs (hunk 3 above)
Comment (KozhanMykhailo) [reply DrawEngine/DrawEngine.cs]: done

File: DrawEngine/DrawEngine.cs (hunk 5 above)
Comment (KozhanMykhailo) [reply DrawEngine/DrawEngine.cs]: done

File: DrawEngine/DrawEngine.cs (hunk 9 above)
Comment (KozhanMykhailo) [reply DrawEngine/DrawEngine.cs]: done

File: Infrastructure/BusinessLogic/BooksWorkflow.cs (hunk 10 above)
Comment (KozhanMykhailo) [reply Infrastructure/BusinessLogic/BooksWorkflow.cs]: done

File: DrawEngine/DrawEngine.cs (hunk 2 above)
Comment (KozhanMykhailo) [reply DrawEngine/DrawEngine.cs]: done

File: DrawEngine/DrawEngine.cs [hunk 11]
```diff
@@ -0,0 +1,555 @@
...
+            dynamicIndent = lengthBook.maxAuthorLength; // Динамічний відступ, який можна змінювати
+            text = "Author";
+            Console.Write(text + new string(' ', dynamicIndent - text.Length) + "|");
+            Console.WriteLine();
+        }
+        public void PrintBookContent(BookLength lengthBook, int id)
+        {
+            var books = _booksWorkflow.GetAll();
+
+            var book = books.SingleOrDefault(s => s.Book_ID == id);
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs:576]: if ==null

File: DrawEngine/DrawEngine.cs [hunk 12]
```diff
@@ -0,0 +1,555 @@
...
+            var book = _booksWorkflow.Get(id);
+            if (book == null)
+            {
+                Console.WriteLine("No record found");
+                PrintBooksMenu();
+                return false;
+            }
+            PrintBook(book);
+            Console.WriteLine("Enter a new book title");
+            string title = Console.ReadLine();
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs:452]: check != string.Empty

File: DrawEngine/DrawEngine.cs [hunk 13]
```diff
@@ -0,0 +1,555 @@
...
+            book.Title = title;
+
+            Console.WriteLine("Enter a new book author");
+            string author = Console.ReadLine();
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs:462]: check != string.Empty

File: DrawEngine/DrawEngine.cs [hunk 14]
```diff
@@ -0,0 +1,555 @@
...
+            if (!int.TryParse(input, out int id)){
+                Console.WriteLine($"The data was entered incorrectly: {input}. Select the menu item");
+                PrintSalesMenu();
+            }
+
+            return _salesWorkflow.DeleteEntity(id);
+        }
+        public bool RemoveBook()
+        {
+            var sales = _salesWorkflow.GetAll();
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: delete

File: DrawEngine/DrawEngine.cs [hunk 15]
```diff
@@ -0,0 +1,555 @@
...
+            else
+                DrawQuestion(number);
+
+            return _salesWorkflow.AddEntity(newSale);
+        }
+        public bool AddBook()
+        {
+            var newBook = new BookDto();
+            Console.WriteLine("Enter the author of the book");
+            newBook.Author = Console.ReadLine();
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: ckeck != string.empty

File: DrawEngine/DrawEngine.cs [hunk 16]
```diff
@@ -0,0 +1,555 @@
...
+            PrintBooksColumnsNames(bookLength);
+            PrintBooksContent(bookLength);
+            PrintBooksLowerBorder(bookLength);
+        }
+
+        public bool AddSale()
+        {
+            var newSale = new SaleDto();
+            Console.WriteLine("Enter the author of the book");
+            newSale.Author = Console.ReadLine();
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: check != string.empty

File: Infrastructure/BusinessLogic/BooksWorkflow.cs [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            return _booksRepository.Get(title, author);
+        }
+
+        public List<Book> GetAll()
+        {
+            return _booksRepository.GetAll();
+        }
+
+        public bool UpdateSaleEntity(Book book)
```
Comment (KozhanMykhailo) [inline Infrastructure/BusinessLogic/BooksWorkflow.cs]: upd name

File: DrawEngine/DrawEngine.cs [hunk 18]
```diff
@@ -0,0 +1,555 @@
...
+                sale.Book_ID = input2;
+
+            }
+            return _salesWorkflow.UpdateSaleEntity(sale);
+        }
+
+        public bool UpdateBook()
+        {
+            Console.WriteLine("Enter the book ID");
+            var idBook = Console.ReadLine();
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs:383]: delete upd book id

File: Infrastructure/Repositories/SalesRepository.cs [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+        public bool Update(Sale sale)
+        {
+            using (ConsoleBookStoreContext context = new ConsoleBookStoreContext())
+            {
+                var existingSale = context.Sales.FirstOrDefault(b => b.Sale_ID == sale.Sale_ID);
+
+                if (existingSale != null)
+                {
+                    existingSale.Price = sale.Price;
```
Comment (KozhanMykhailo) [inline Infrastructure/Repositories/SalesRepository.cs:63]: add upd Book_Id

File: DrawEngine/DrawEngine.cs [hunk 20]
```diff
@@ -0,0 +1,555 @@
...
+                sale.Number_Of_Sales = result1;
+            else
+                Console.WriteLine();
+
+            Console.WriteLine("Сhange ID of book?  1 - yes, 2 - no");
+            var input1 = ReadEnteredValue();
+            if (input1 == 1)
+            {
+                Console.WriteLine("Enter a new book ID");
+                var input2 = ReadEnteredValue();
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: check if book exists

File: DrawEngine/DrawEngine.cs [hunk 21]
```diff
@@ -0,0 +1,555 @@
...
+            PrintBook(book);
+            Console.WriteLine("Enter a new book title");
+            string title = Console.ReadLine();
+            book.Title = title;
+
+            Console.WriteLine("Enter a new book author");
+            string author = Console.ReadLine();
+            book.Author = author;
+
+            Console.WriteLine("Change ID of book?  1 - yes, 2 - no");
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: delete this part

File: DrawEngine/DrawEngine.cs [hunk 22]
```diff
@@ -0,0 +1,555 @@
...
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex.Message);
+                return -1;
+            }
+        }
+
+        public SaleLength CalculateSalesLength() 
+        {
+            var sales = _salesWorkflow.GetAll();
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs:66]: if sales.count == 0 ->  return new SaleLength {maxSaleIDLength = "Sale ID".lenth , maxTitleLength ="Title".lenth ]

File: DrawEngine/DrawEngine.cs [hunk 23]
```diff
@@ -0,0 +1,555 @@
...
+            str = "Number Of Sales";
+            lengthSales.maxNumberOfSalesLength = sales.Max(s => s.Number_Of_Sales.ToString().Length) > str.Length ? sales.Max(s => s.Number_Of_Sales.ToString().Length) : str.Length;
+
+            lengthSales.maxStringLength = lengthSales.maxSaleIDLength + lengthSales.maxTitleLength + lengthSales.maxAuthorLength + lengthSales.maxPriceLength + lengthSales.maxNumberOfSalesLength;
+            return lengthSales;
+        }
+        public BookLength CalculateBookLength()
+        {
+            var books = _booksWorkflow.GetAll();
+
```
Comment (KozhanMykhailo) [inline DrawEngine/DrawEngine.cs]: if boos == 0 -> ...

File: DrawEngine/DrawEngine.cs (hunk 11 above)
Comment (KozhanSerhii) [reply DrawEngine/DrawEngine.cs:576]: done

File: DrawEngine/DrawEngine.cs (hunk 12 above)
Comment (KozhanSerhii) [reply DrawEngine/DrawEngine.cs:400]: done

Comment (KozhanSerhii) [reply DrawEngine/DrawEngine.cs:452]: done

File: DrawEngine/DrawEngine.cs (hunk 13 above)
Comment (KozhanSerhii) [reply DrawEngine/DrawEngine.cs:462]: done

File: DrawEngine/DrawEngine.cs (hunk 14 above)
Comment (KozhanSerhii) [reply DrawEngine/DrawEngine.cs]: done

File: DrawEngine/DrawEngine.cs (hunk 15 above)
Comment (KozhanSerhii) [reply DrawEngine/DrawEngine.cs]: done

File: DrawEngine/DrawEngine.cs (hunk 16 above)
Comment (KozhanSerhii) [reply DrawEngine/DrawEngine.cs]: done

File: Infrastructure/BusinessLogic/BooksWorkflow.cs (hunk 17 above)
Comment (KozhanSerhii) [reply Infrastructure/BusinessLogic/BooksWorkflow.cs]: done

File: Infrastructure/Repositories/SalesRepository.cs (hunk 19 above)
Comment (KozhanSerhii) [reply Infrastructure/Repositories/SalesRepository.cs:63]: done

File: DrawEngine/DrawEngine.cs (hunk 20 above)
Comment (KozhanSerhii) [reply DrawEngine/DrawEngine.cs]: done

File: DrawEngine/DrawEngine.cs (hunk 21 above)
Comment (KozhanSerhii) [reply DrawEngine/DrawEngine.cs]: done

File: DrawEngine/DrawEngine.cs (hunk 23 above)
Comment (KozhanSerhii) [reply DrawEngine/DrawEngine.cs]: done

File: DrawEngine/DrawEngine.cs (hunk 22 above)
Comment (KozhanSerhii) [reply DrawEngine/DrawEngine.cs:66]: done

File: Infrastructure/Repositories/SalesRepository.cs [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+                    {
+                        Console.WriteLine("Enter a new book ID");
+                        if (long.TryParse(Console.ReadLine(), out long result))
+                        {
+                            var book = context.Books.SingleOrDefault(b => b.Book_ID == result);
+                            if (book != null)
+                                sale.Book_ID = result;
+                            else
+                                Console.WriteLine("The book with this ID does not exist");
+                        }
```
Comment (KozhanMykhailo) [inline Infrastructure/Repositories/SalesRepository.cs]: else {cw(new id has inccorect format, won't be changed)}

File: Infrastructure/Repositories/SalesRepository.cs (hunk 19 above)
Comment (KozhanMykhailo) [reply Infrastructure/Repositories/SalesRepository.cs:63]: тут треба було просто написати це - sale.Book_ID = result;
все інше винести треба назад в drawEngine

**Comment by KozhanSerhii** [comment]
!o

**Pull request merged**